kratos/utilities/variable_utils.h

```diff
@@ -504,6 +504,31 @@ class KRATOS_API(KRATOS_CORE) VariableUtils
         KRATOS_CATCH("")
     }
 
+    /**
+     * @brief Distributes variable values in TContainerType container to nodes
+     *
+     * This method distributes variables values stored in TContainerType data value container in rModelPart
+     * to nodes. Constant weighting is used for each node based on rWeightVariable value. The result
+     * is stored in nodal non-historical data value container under the same rVariable. If IsInverseWeightProvided
+     * is true, then the weights provided by rWeightVariable is inverted to get nodal weight. Otherwise, the value
+     * given by rWeightVariable is used as weight.
+     *
+     *
+     * @tparam TDataType               Data type
+     * @tparam TContainerType          ContainerType of model part
+     * @tparam TWeightDataType         Data type of weight variable (this should be either int or double)
+     * @param rModelPart               Model part
+     * @param rVariable                Variable to be distributed
+     * @param rWeightVariable          Variable which holds weight to distribute entity values to nodes
+     * @param IsInverseWeightProvided  Whether the weight is provided as inverse or not.
+     */
+    template <class TDataType, class TContainerType, class TWeightDataType>
+    void DistributeVariable(
+        ModelPart& rModelPart,
+        const Variable<TDataType>& rVariable,
+        const Variable<TWeightDataType>& rWeightVariable,
```

Comment:
In which cases the distribution variable can be a non-double variable?

Reply:
The problem is with integers. Since this supports taking the inverse weighting given by the variable, so one can use MPI assembly techniques to properly assemble integers (such as neighbour conditions, etc.), So I though it wouldn't harm anyone to have it :)

One other reason is, In order to have #7039 working I need to use NUMBER_OF_NEIGHBOUR_CONDITIONS. So if I only have double version of TWeightDataType, then I have to define NUMBER_OF_NEIGHBOUR_CONDITIONS as double which contradicts with RANSApplication same variable defined as integer.

Reply:
I was thinking in something like this. The point is if this has physical meaning. I'd expect this operation to be always weighted by the area/volume of the condition/elements.

Furthermore, I'm wondering if this somehow contradicts what is implemented in the `VariableRedistributionUtils` which was designed to be conservative.

Reply:
> I was thinking in something like this. The point is if this has physical meaning. I'd expect this operation to be always weighted by the area/volume of the condition/elements.

Yes that is correct, it will be weighted by always area/volume/number_of_neighbours of the conditions/elements. So inverse makes sense most of the time. For me also inverse works, but I was thinking of being general as possible to support both inverse and direct.

One other use case which I already see (and which is not supported by these methods) is, first getting the conditional variable, and devide it by number of nodes in condition and then add it to all the nodes. Example is, lets say we want to calculate wall shear stress where SLIP and wall laws are used. So first we have to calculate u_tau at wall condition, then calculate tau, then multiply it by condition surface area to get total shear force acting on the conditon surface (this is stored as a vector variable in conditional data value container, say SHEAR_FORCE variable). Finally we need to distribute it among nodes, since drag calculation processes needs to have **NODAL REACTION** properly calculated. So we have to devide conditional SHEAR_FORCE vector by number of nodes in the condition, and add it to the REACTION variable of the nodes so once you accumulate all the nodal REACTIONS, it reflects total REEACTION as in the accumulated CONDITIONAL SHEAR_FORCE. (I already have to do this in RANSApplication, since I always assume SLIP walls with wall laws for RANS)

I was thinking of implementing this as well, then I came accross the problem with naming, I couldn't come up with proper naming so user won't get confused.

> Furthermore, I'm wondering if this somehow contradicts what is implemented in the VariableRedistributionUtils which was designed to be conservative.

Sorry, this I don't understand. :/

Reply:
> > I was thinking in something like this. The point is if this has physical meaning. I'd expect this operation to be always weighted by the area/volume of the condition/elements.
> 
> Yes that is correct, it will be weighted by always area/volume/number_of_neighbours of the conditions/elements. So inverse makes sense most of the time. For me also inverse works, but I was thinking of being general as possible to support both inverse and direct.

OK. Let's keep it then.
> 
> One other use case which I already see (and which is not supported by these methods) is, first getting the conditional variable, and devide it by number of nodes in condition and then add it to all the nodes. Example is, lets say we want to calculate wall shear stress where SLIP and wall laws are used. So first we have to calculate u_tau at wall condition, then calculate tau, then multiply it by condition surface area to get total shear force acting on the conditon surface (this is stored as a vector variable in conditional data value container, say SHEAR_FORCE variable). Finally we need to distribute it among nodes, since drag calculation processes needs to have **NODAL REACTION** properly calculated. So we have to devide conditional SHEAR_FORCE vector by number of nodes in the condition, and add it to the REACTION variable of the nodes so once you accumulate all the nodal REACTIONS, it reflects total REEACTION as in the accumulated CONDITIONAL SHEAR_FORCE. (I already have to do this in RANSApplication, since I always assume SLIP walls with wall laws for RANS)
> 
This is possible an independent discussion, but I take the chance to say that the reaction-based calculation possibly is not the best way to proceed, mainly because of the wall laws as you point out.
Instead I'd do it as in the embedded case, implementing a `DRAG_FORCE` variable calculation in all the conditions and save the values in the Gauss points. This is basically the same but keeping the values as a Gauss point variable as it is equivalent to a structural mechanics stress.

> I was thinking of implementing this as well, then I came accross the problem with naming, I couldn't come up with proper naming so user won't get confused.
> 
> > Furthermore, I'm wondering if this somehow contradicts what is implemented in the VariableRedistributionUtils which was designed to be conservative.
> 
> Sorry, this I don't understand. :/

Forget about this, I was thinking in a way different application (converting a nodal value to a distributed one). The point is that the usage of these might be misunderstood with the `VariableRedistribuionUtility` methods. Don't you thing that `ConcentrateVariable` or `AccumulateVariableOnNodes` would be a better name?

Reply:
> This is possible an independent discussion, but I take the chance to say that the reaction-based calculation possibly is not the best way to proceed, mainly because of the wall laws as you point out.
Instead I'd do it as in the embedded case, implementing a DRAG_FORCE variable calculation in all the conditions and save the values in the Gauss points. This is basically the same but keeping the values as a Gauss point variable as it is equivalent to a structural mechanics stress.

I will change the way I do drag calculation in RANS. I can follow the way you do in embedded and I also see that as the most apprpriate way to calculate it :)

>This is possible an independent discussion, but I take the chance to say that the reaction-based calculation possibly is not the best way to proceed, mainly because of the wall laws as you point out.
Instead I'd do it as in the embedded case, implementing a DRAG_FORCE variable calculation in all the conditions and save the values in the Gauss points. This is basically the same but keeping the values as a Gauss point variable as it is equivalent to a structural mechanics stress.

What do you say to `WeightedAccumulateEntityVariableOnNodes` ?

Reply:
> > This is possible an independent discussion, but I take the chance to say that the reaction-based calculation possibly is not the best way to proceed, mainly because of the wall laws as you point out.
> > Instead I'd do it as in the embedded case, implementing a DRAG_FORCE variable calculation in all the conditions and save the values in the Gauss points. This is basically the same but keeping the values as a Gauss point variable as it is equivalent to a structural mechanics stress.
> 
> I will change the way I do drag calculation in RANS. I can follow the way you do in embedded and I also see that as the most apprpriate way to calculate it :)
> 
OK. We can discuss about it once you have a look to it.

> > Forget about this, I was thinking in a way different application (converting a nodal value to a distributed one). The point is that the usage of these might be misunderstood with the VariableRedistribuionUtility methods. Don't you thing that ConcentrateVariable or AccumulateVariableOnNodes would be a better name?
> 
> What do you say to `WeightedAccumulateEntityVariableOnNodes` ?

I like `WeightedAccumulateVariableOnNodes`. :wink:

Reply:
done :)